backends/platform/ios7/ios7_video_metal.mm

```diff
@@ -0,0 +1,292 @@
+		_openGLContext = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2 sharegroup:_mainContext.sharegroup];
+
+		if (_openGLContext == nil) {
+			printError("Could not create OpenGL ES context using sharegroup");
+			abort();
+		}
+		// background thread will always use _openGLContext
+		if ([EAGLContext setCurrentContext:_openGLContext]) {
+			[self setupRenderBuffer];
+		}
+	}
+	return CVOpenGLESTextureGetName(_openGLTextureRef);
+}
+
+- (void)destroyOpenGLContext {
+	[_openGLContext release];
+	_openGLContext = nil;
+}
+
+- (void)setupOpenGLTextureCache {
+	if (_openGLTextureCache == nullptr) {
+		CVOpenGLESTextureCacheCreate(
+			kCFAllocatorDefault,
+			NULL, _openGLContext,
+			NULL, &_openGLTextureCache);
+	}
+
+	CFDictionaryRef emptyDict = CFDictionaryCreate(
+		kCFAllocatorDefault,
+		NULL, NULL, 0,
+		&kCFTypeDictionaryKeyCallBacks,
+		&kCFTypeDictionaryValueCallBacks);
+
+	CFMutableDictionaryRef attributes = CFDictionaryCreateMutable(
+		kCFAllocatorDefault, 1,
+		&kCFTypeDictionaryKeyCallBacks,
+		&kCFTypeDictionaryValueCallBacks);
+
+	CFDictionarySetValue(
+		attributes,
+		kCVPixelBufferIOSurfacePropertiesKey, // This creates a shared IOSSurface
+		emptyDict);
+
+	if (_sharedPixelBuffer != nullptr) {
+		CFRelease(IOSurfaceRef(_sharedPixelBuffer));
+		_sharedPixelBuffer = nullptr;
+	}
+
+	CVPixelBufferCreate(
+		kCFAllocatorDefault,
+		_drawableWidth, _drawableHeight,
+		kCVPixelFormatType_32BGRA,
+		attributes,
+		&_sharedPixelBuffer);
+
+	CFRelease(attributes);
+	CFRelease(emptyDict);
+
+	if (_openGLTextureRef != nullptr) {
+		CVBufferRelease(_openGLTextureRef);
+	}
+	CVOpenGLESTextureCacheFlush(_openGLTextureCache, 0);
+
+	CVOpenGLESTextureCacheCreateTextureFromImage(
+		kCFAllocatorDefault,
+		_openGLTextureCache,
+		_sharedPixelBuffer,
+		NULL,
+		GL_TEXTURE_2D,
+		GL_RGBA,
+		_drawableWidth,
+		_drawableHeight,
+		GL_RGBA,
+		GL_UNSIGNED_BYTE,
+		0,
+		&_openGLTextureRef);
+}
+
+- (void)setupMetalTextureCache {
+	if (_metalTextureCache == nullptr) {
+		CVMetalTextureCacheCreate(
+			kCFAllocatorDefault,
+			nil,
+			_metalDevice,
+			nil,
+			&_metalTextureCache);
+	}
+
+	if (_metalTextureRef != nullptr) {
+		CVBufferRelease(_metalTextureRef);
+	}
+	CVMetalTextureCacheFlush(_metalTextureCache, 0);
+
+	CVMetalTextureCacheCreateTextureFromImage(
+		kCFAllocatorDefault,
+		_metalTextureCache,
+		_sharedPixelBuffer,
+		NULL,
+		MTLPixelFormatRGBA8Unorm,
+		_drawableWidth, _drawableHeight,
+		0,
+		&_metalTextureRef);
+
+	_metalTexture = CVMetalTextureGetTexture(_metalTextureRef);
+}
+
+- (void)refreshScreen {
+	NSError *error = NULL;
+
+	glFinish();
+
+	@autoreleasepool {
+		id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];
+		id<CAMetalDrawable> drawable = [_metalLayer nextDrawable];
+		id<MTLComputeCommandEncoder> encoder = [commandBuffer computeCommandEncoder];
+
+		[encoder setComputePipelineState:[_metalDevice newComputePipelineStateWithFunction:_kernelFunctionFlipY error:&error]];
```

Comment:
I suppose this should use CVOpenGLESTextureIsFlipped to flip only when it's needed?